src/librustc_mir/interpret/eval_context.rs

```diff
@@ -546,57 +566,143 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'a, 'mir, 'tcx>> InterpretCx<'a, 'mir, 'tc
-                // Data got changed, better make sure it matches the type!
-                // It is still possible that the return place held invalid data while
-                // the function is running, but that's okay because nobody could have
-                // accessed that same data from the "outside" to observe any broken
-                // invariant -- that is, unless a function somehow has a ptr to
-                // its return place... but the way MIR is currently generated, the
-                // return place is always a local and then this cannot happen.
-                self.validate_operand(
-                    self.place_to_op(return_place)?,
-                    vec![],
-                    None,
-                    /*const_mode*/false,
-                )?;
+
+
+        let do_normal_ret = self.unwinding != Unwinding { in_cleanup: false };
+        trace!("unwind: performing normal return: {}", do_normal_ret);
+        // When we're popping a stack frame for unwinding purposes,
+        // we don't care at all about returning-related stuff (i.e. return_place
+        // and return_to_block), because we're not performing a return from this frame.
+        if do_normal_ret  {
+            // Validate the return value. Do this after deallocating so that we catch dangling
+            // references.
+            if let Some(return_place) = frame.return_place {
+                if M::enforce_validity(self) {
+                    // Data got changed, better make sure it matches the type!
+                    // It is still possible that the return place held invalid data while
+                    // the function is running, but that's okay because nobody could have
+                    // accessed that same data from the "outside" to observe any broken
+                    // invariant -- that is, unless a function somehow has a ptr to
+                    // its return place... but the way MIR is currently generated, the
+                    // return place is always a local and then this cannot happen.
+                    self.validate_operand(
+                        self.place_to_op(return_place)?,
+                        vec![],
+                        None,
+                        /*const_mode*/false,
+                    )?;
+                }
+            } else {
+                // Uh, that shouldn't happen... the function did not intend to return
+                return err!(Unreachable);
             }
-        } else {
-            // Uh, that shouldn't happen... the function did not intend to return
-            return err!(Unreachable);
+
+
+            // Jump to new block -- *after* validation so that the spans make more sense.
+            match frame.return_to_block {
+                StackPopCleanup::Goto { ret, ..  } => {
+                    self.goto_block(ret)?;
+                }
+                StackPopCleanup::None { .. } => {}
+            }
+
         }
```

Comment:
Why is there no `else` here that jumps to the unwind block?

The logic should be something like
```rust
let next_block = if do_normal_ret {
  // First perform return value validation.
  // Now determine next block for normal return.
  match frame.return_to_block {
    Goto { ret, .. } => Some(ret),
    None { .. } => None
  }
} else {
  // Determine next block for unwinding return.
  match frame.return_to_block {
    Goto { unwind, .. } => Some(unwind),
    None => bug!("Unwinding across special-return frame, that should not be possible"),
  }
};
if let Some(block) = next_block { self.goto_block(block)?; }
```
I guess the `goto_block` could also be embedded inside the first conditional, though I think this structure shows the uniformity better.

But it seems you moved the `goto_block` in case of unwinding elsewhere? I think that just obscures the control flow, I wouldn't do that. If you want to be *really sure* that there is no unwinding during CTFE, add a `const UNWINDING_SUPPORTED: bool` to the machine and assert that it is `true` in the unwinding case here.

Reply:
I deliberately moved jumping to the unwind block into `pop_stack_frame`.
If I were to combine `pop_stack_frame` and `pop_stack_frame_internal` back into `pop_stack_frame`, it would need to call itself recursively in order to properly pop the stack. That would make the logic incredibly hard to follow, so I opted to split it into two methods.

Reply:
> I deliberately moved jumping to the unwind block into pop_stack_frame.

Please don't. Jumping to the unwind block and jumping to the "normal return" block should be handled as symmetrically as possible. They are very similar operations.

>  it would need to call itself recursively in order to properly pop the stack.

Why that? I see no need for this.

Reply:
`pop_stack_frame` needs to handle the case where we have multiple 'empty' frames (e.g. frames without a cleanup block) between the current frame and the next non-'empty' frame. This requires repeatedly popping frames from the stack. Splitting `pop_stack_frame` into two methods allows the normal popping logic to be separated out, allowing us to pop multiple frames at once without recursion.

Reply:
> pop_stack_frame needs to handle the case where we have multiple 'empty' frames (e.g. frames without a cleanup block) between the current frame and the next non-'empty' frame.

When does that happen?
If, during unwinding, we encounter a frame without `cleanup` continuation, isn't that just a bug?

Reply:
No - not all frames need cleanup. In that case, we simply pop the frame and move on to the next one.

Reply:
Interesting, so MIR has some kind of built-in "bubble-up" semantics... this is demonstrated by e.g. the MIR for the call in
```rust
fn foo(f: fn(i32) -> i32) -> i32 {
    f(42)
}
```
being `_0 = move _2(const 42i32) -> bb1`. No unwind edge. That implicitly uses the unwind edge of `foo` then.

Doing all these stack pops in a single machine step seems odd though. I wonder if it would be nicer to have a way to represent the intermediate states here? Not sure what a good way to do that would be, though.

Reply:
I considered that - however, it would require a large refactor of how we handle stepping through frames. We would need to keep track of which frames have actually 'finished' (e.g. should be popped despite still pointing to the terminator). I think that would simply make the logic harder to follow, without exposing any useful information (after all, nothing is actually happening when we pop the intermediate frames).

Reply:
> after all, nothing is actually happening when we pop the intermediate frames

That's not correct, we are freeing tons of memory.

> however, it would require a large refactor of how we handle stepping through frames. We would need to keep track of which frames have actually 'finished' (e.g. should be popped despite still pointing to the terminator)

I don't think it would be a big refactor. I think all we need is for a stack frame to be able to have "no" current basic block + statement. Then `step` would interpret that the same way as an `Unwind` terminator, and `pop_stack_frame`, when unwinding a frame that has no unwind continuation, would set the caller's next block/statement to "none (unwinding)".

Reply:
Conceptually, making all these pops a single "machine step" does not seem right, so IMO if avoiding it is not too much of an effort, we should avoid it. Generally, loops in a single interpreter step should be a warning sign -- we need them to iterate over all the bytes loaded/stored from memory, or to evaluate a `Place` which is a recursive structure, but that should be about it in terms of loops inside the main interpreter loop.

Reply:
> Conceptually, making all these pops a single "machine step" does not seem right

I'm not sure that I agree. The only thing that we do for these empty frames is deallocating memory. This is not observable by the user - none of the locals can have any drop glue, since we would otherwise have an unwind edge. This is essentially just collapsing several intermediate steps into one.

> I don't think it would be a big refactor. I think all we need is for a stack frame to be able to have "no" current basic block + statement. 

That feels very weird to me, especially considering that this would never be used outside of unwinding. I don't really like the idea of adding a new state that doesn't correspond to anything the MIR or (I believe) anything that the panic runtime does.

Reply:
>  This is essentially just collapsing several intermediate steps into one.

I see that. Every time one writes an interpreter there's some places where one can split steps or merge them, and there's no clear best answer.

But in my experience, it is best to leave loops out of the interpreter as much as we can -- to make [this](https://github.com/rust-lang/rust/blob/857a55b8f0a692cb2304ab7c062d9cc74122e516/src/librustc_mir/interpret/step.rs#L40) the only loop that we have. Now, we already have other loops to iterate over bytes in memory when handling arbitrary-sized data, but there I think it is clear that making these all single interpreter steps would not make any sense.

> I don't really like the idea of adding a new state that doesn't correspond to anything the MIR

The MIR is a data structure for holding code, I don't see how *any* state we have corresponds to anything there.

OTOH I mentioned several arguments for not having a loop here:
* It simplifies the "pop frame" code, in particular it helps expose the concept that "return" and "unwind" are just two ways to exit a function. I think this helps a lot in understanding unwinding at a conceptual level, and it'd be good to have that symmetry reflected in the code. But currently it isn't, which is what started this discussion. I find it harder to review this code because of that loop; it makes it much more complicated to to relate the simple conceptual model to what you implemented.
* There exists a Miri frontend that makes the individual steps user-visible, [Priroda](https://github.com/oli-obk/priroda/). The bigger the steps, the less well that works and the more "magic" things look to the user.

You claimed a big refactor was necessary to change this; I do not think so. Maybe I should try doing it? That would probably also help me see if everything else you did here makes sense to me.

The reason I care so much about what might seem like details to you is that this PR is the biggest conceptual extension of the Miri engine since I joined. I would like to be sure that it actually matches my mental model if what I think it should be (I had plans for implementing this but just no time to do it), and when it doesn't I need to understand why.

Reply:
Okay, I wrote a patch at https://github.com/RalfJung/rust/commit/610903f5af26cd9222d1306f9fd4c4b2c14eea0c. It compiles but I didn't test it. The net line diff is negative even though this also adds some extra sanity checks (like only accepting `Resume` inside a cleanup block). In fact, if [this assertion](https://github.com/RalfJung/rust/commit/610903f5af26cd9222d1306f9fd4c4b2c14eea0c#diff-470c35c1f39f9d3f8817435a63c73ad2R572) passes that demonstrates that the `unwind` argument to `pop_stack_frame` is redundant -- though I'd still be in favor of keeping it as a sanity check.

Reply:
Personally, I still it's clearer to jump through several cleanup steps at once. However, I'm okay going with your approach, and I've applied your patch.

Reply:
Thanks!
Does this mean the assertion passes, or should I run some tests?